File: 07/07.py
from collections import defaultdict

from timing_util import Timing


def get_data(content: str) -> dict[str, int]:
    sizes = defaultdict(int)
    cwd = ['']

    for line in content.splitlines():
        match line.removesuffix('\n').split(' '):
            case ['$', 'cd', '/']:
                cwd = ['/']
            case ['$', 'cd', '..']:
                cwd = cwd[:-1]
            case ['$', 'cd', target]:
                cwd.append(target)
            case [num, _] if num.isdigit():
                s = int(num)

                pwd = '/'
                sizes[pwd] += s
                for part in cwd:
                    pwd += part + '/'
                    sizes[pwd] += s

    return sizes


def part1(sizes: dict[str, int]) -> int:
    return sum(s for s in sizes.values() if s <= 100e3)


def part2(sizes: dict[str, int]) -> int:
    min_free = sizes['/'] - (70e6 - 30e6)
    return min(s for s in sizes.values() if s >= min_free)


if __name__ == '__main__':
    with Timing():
        with open('07/input.txt') as in_file:
            sizes = get_data(in_file.read())

        print(f'part1: {part1(sizes)}')
        print(f'part2: {part2(sizes)}')

File: 07/07.rs
use std::fs;
use std::collections::HashMap;
use std::time::Instant;

fn main() {
    let start = Instant::now();

    let sizes: HashMap<String, usize> = file_system("07/input.txt");

    let p1: usize = sizes.values().filter(|size| **size <= 100_000).sum();
    println!("part1: {}", p1);

    let free_size: usize = sizes["/"] - (70_000_000 - 30_000_000);
    let p2: usize = *sizes.values().filter(|size| **size >= free_size).min().unwrap();
    println!("part2: {}", p2);

    println!("time: {:?}", start.elapsed());
}

fn file_system(path: &str) -> HashMap<String, usize> {
    let data: String = fs::read_to_string(path).expect("File not found");
    let mut sizes: HashMap<String, usize> = HashMap::new();
    let mut cwd: Vec<String> = vec!["".to_string()];

    for line in data.split("\n") {
        let parts: Vec<&str> = line.split(" ").collect();
        match parts[0..2] {
            ["$", "cd"] => {
                match parts[2] {
                    "/" => {
                        cwd = vec!["/".to_string()];
                    },
                    ".." => {
                        cwd.pop();
                    },
                    _ => {
                        cwd.push(parts[2].to_string());
                    }
                }
            },
            [_, _] => {
                let num = parts[0].parse::<usize>();
                match num {
                    Ok(val) => {
                        let mut dir = "/".to_string();
                        *sizes.entry(dir.to_string()).or_insert(0) += val;

                        // update size in cwd and all parents
                        for part in cwd.iter().skip(1) {
                            dir.push_str(part);
                            dir.push_str("/");
                            *sizes.entry(dir.to_string()).or_insert(0) += val;
                        }
                    },
                    _ => {}
                }
            },
            _ => {}
        }
    }
    return sizes
}
